sklearn/utils/sparsefuncs_fast.pyx

```diff
@@ -27,27 +27,35 @@ def csr_row_norms(X):
     """L2 norm of each row in CSR matrix X."""
     if X.dtype not in [np.float32, np.float64]:
         X = X.astype(np.float64)
-    return _csr_row_norms(X.data, X.shape, X.indices, X.indptr)
+    return np.asarray(_csr_row_norms(X.data, X.indices, X.indptr))
 
 
-def _csr_row_norms(cnp.ndarray[floating, ndim=1, mode="c"] X_data,
-                   shape,
-                   cnp.ndarray[integral, ndim=1, mode="c"] X_indices,
-                   cnp.ndarray[integral, ndim=1, mode="c"] X_indptr):
+def _csr_row_norms(
+    floating[::1] X_data,
+    integral[::1] X_indices,
+    integral[::1] X_indptr,
+):
     cdef:
-        unsigned long long n_samples = shape[0]
-        unsigned long long i
+        integral n_samples = X_indptr.shape[0] - 1
+        integral i
         integral j
-        double sum_
+        floating sum_
 
-    norms = np.empty(n_samples, dtype=X_data.dtype)
-    cdef floating[::1] norms_view = norms
+    if floating is float:
+        dtype = np.float32
+    else:
+        dtype = np.float64
 
-    for i in range(n_samples):
-        sum_ = 0.0
-        for j in range(X_indptr[i], X_indptr[i + 1]):
-            sum_ += X_data[j] * X_data[j]
-        norms_view[i] = sum_
+    cdef floating[::1] norms = (
+        np.empty(n_samples, dtype=dtype)
+    )
+
+    with nogil:
+        for i in range(n_samples):
+            sum_ = 0.0
+            for j in range(X_indptr[i], X_indptr[i + 1]):
+                sum_ += X_data[j] * X_data[j]
+            norms[i] = sum_
```

Comment:
If we initialize `norms` to zeros instead, would the following be much slower?

```suggestion
    with nogil:
        for i in range(n_samples):
            for j in range(X_indptr[i], X_indptr[i + 1]):
                norms[i] += X_data[j] * X_data[j]
```

I see that here you have to access every time the element in the array to accumulate the sum but I am not sure of the overhead.

Reply:
`np.empty` is faster than `np.zeros`, yet we should be able to initialise each element of `norms` to 0.0 and not use `sum_`: 
```suggestion
    with nogil:
        for i in range(n_samples):
            norms[i] = 0.0
            for j in range(X_indptr[i], X_indptr[i + 1]):
                norms[i] += X_data[j] * X_data[j]
```

Reply:
> np.empty is faster than np.zeros

I think that here we should check if `np.empty` + a for loop initializing at 0 is faster than `np.zeros`. I am kind of surprised that this is actually faster since I would expect to do exactly this at the C level. Is there any extra overhead involved in the `np.zeros` operation?

Reply:
I would expect that np.zeros is faster because it should be a malloc + memset which is faster than a malloc + loop

Reply:
We should probably benchmarks both alternatives.